plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -182,35 +207,157 @@ public void migrate(ConnectorSession session, String schemaName, String tableNam
+        else if (!"MANAGED_TABLE".equalsIgnoreCase(hiveTable.getTableType()) && !"EXTERNAL_TABLE".equalsIgnoreCase(hiveTable.getTableType())) {
             throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support migrating %s table type".formatted(hiveTable.getTableType()));
         }
-        if (isDeltaLakeTable(hiveTable)) {
-            throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support migrating Delta Lake tables");
+        else if (isDeltaLakeTable(hiveTable)) {
+            migrateDeltaLakeTable(session, metastore, sourceTableName, hiveTable);
         }
-        if (isHudiTable(hiveTable)) {
+        else if (isHudiTable(hiveTable)) {
             throw new TrinoException(NOT_SUPPORTED, "The procedure doesn't support migrating Hudi tables");
         }
-        if (isIcebergTable(hiveTable)) {
+        else if (isIcebergTable(hiveTable)) {
             throw new TrinoException(NOT_SUPPORTED, "The table is already an Iceberg table");
         }
+        else {
+            RecursiveDirectory recursive = Enums.getIfPresent(RecursiveDirectory.class, recursiveDirectory.toUpperCase(ENGLISH)).toJavaUtil()
+                    .orElseThrow(() -> new TrinoException(INVALID_PROCEDURE_ARGUMENT, "Invalid recursive_directory: " + recursiveDirectory));
+            migrateHiveTable(session, recursive, metastore, sourceTableName, hiveTable);
+        }
+    }
+
+    private void migrateDeltaLakeTable(
+            ConnectorSession session,
+            HiveMetastore metastore,
+            SchemaTableName schemaTableName,
+            io.trino.plugin.hive.metastore.Table hiveTable)
+    {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        TrinoCatalog catalog = catalogFactory.create(session.getIdentity());
+
+        Map<String, String> serdeParameters = hiveTable.getStorage().getSerdeParameters();
+        String location = serdeParameters.get(PATH_PROPERTY);
+        if (location == null) {
+            throw new TrinoException(DELTA_LAKE_INVALID_SCHEMA, format("No %s property defined for table: %s", PATH_PROPERTY, hiveTable));
+        }
+
+        TableSnapshot tableSnapshot;
+        try {
+            tableSnapshot = transactionLogAccess.loadSnapshot(schemaTableName, location, session);
+        }
+        catch (IOException e) {
+            throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Unable to load Delta Lake table for migration", e);
+        }
+
+        ProtocolEntry protocolEntry = transactionLogAccess.getProtocolEntries(tableSnapshot, session)
+                .reduce((first, second) -> second)
+                .orElseThrow(() -> new TrinoException(DELTA_LAKE_INVALID_SCHEMA, "Protocol entry not found in transaction log for table " + tableSnapshot.getTable()));
+        MetadataEntry metadataEntry = transactionLogAccess.getMetadataEntry(tableSnapshot, session);
+
+        if (protocolEntry.getMinReaderVersion() > 1) {
```

Comment:
These constraints are to avoid dealing with column mapping modes correct? https://docs.databricks.com/delta/delta-column-mapping.html

Reply:
That's right. I think we can deal with column mapping modes, but I wanted to start with the simple case first